src/main.ts

```diff
@@ -341,103 +319,87 @@ export class BotFrameworkInstrumentation {
-    let item = {
+  trackCustomEvent(session: builder.Session, eventName: string, customProperties: IDictionary) {
+    this.logEvent(session, eventName, customProperties);
+  }
+
+  private getLogProperties(session: builder.Session | builder.IMessage, properties?: IDictionary): any {
+
+    if (session == null) { return properties || null; }
+
+    let message: builder.IMessage | any = {};
+    let isSession = false;
+
+    if ((<any>session).message) {
+      isSession = true;
+      message = (<any>session).message;
+    } else {
+      message = session;
+    }
+
+    let address: builder.IAddress | any = message.address || {};
+    let conversation: builder.IIdentity = address.conversation || {};
+    let user: builder.IIdentity = address.user || {};
+    let item: any = {
       timestamp: message.timestamp,
       channel: address.channelId,
       conversationId: conversation.id,
-      userId: user.id,
-      userName: user.name
+      userId: user.id
     };
-    //merge the custom properties with the defaults
-    let eventData = Object.assign(item, keyValuePair);
-    this.trackEvent(eventName, eventData);
+
+    if (!this.settings.omitUserName) {
+      item.userName = user.name;
+    }
+
+    // Adding custom fields if present
+    if (isSession && this.customFields) {
+      PROPERTY_BAGS.forEach(propertyBag => {
+        let properties = this.customFields[propertyBag] || [];
+        properties.forEach(property => {
+          item[property] = session[propertyBag][property] || null;
+        });
+      });
+    }
+
+    return Object.assign(item, properties);
   }
 
   /**
    * Log a user action or other occurrence.
    * @param name              A string to identify this event in the portal.
    * @param properties        map[string, string] - additional data used to filter events and metrics in the portal. Defaults to empty.
-   * @param measurements      map[string, number] - metrics associated with this event, displayed in Metrics Explorer on the portal. Defaults to empty.
-   * @param tagOverrides      the context tags to use for this telemetry which overwrite default context values
-   * @param contextObjects    map[string, contextObject] - An event-specific context that will be passed to telemetry processors handling this event before it is sent. For a context spanning your entire operation, consider appInsights.getCorrelationContext
    */
-  private trackEvent(
-    name: string, 
-    properties?: {[key: string]: string;}, 
-    measurements?: {[key: string]: number;}, 
-    tagOverrides?: {[key: string]: string;}, 
-    contextObjects?: {[name: string]: any;}): void   {
-    _.forEach(this.appInsightsClients, (client) => {
-      client.trackEvent(name, properties);
-    });
+  private logEvent(session: builder.Session, name: string, properties?: IDictionary): void   {
+    this.appInsightsClients.forEach(client => client.trackEvent(name, this.getLogProperties(session, properties)));
```

Comment:
@morsh 
The code is repeatedly being called in each loop which makes it inefficient. 
I recommend to extract `this.getLogProperties(session, properties)` before the loop:
```
private logEvent(session: builder.Session, name: string, properties?: IDictionary): void   {
   let logProps =  this.getLogProperties(session, properties);
   this.appInsightsClients.forEach(client => client.trackEvent(name, logProps));
}
```